### ISSUE TITLE
Editor ignores backspace or delete

### ISSUE THREAD

**Issue by paplorinc** [cursiveclojure/cursive#788]
Hello, I was playing with the editor to try it out and noticed that in some cases the backspace or delete key doesn't do anything.

E.g. in the following code example:
```clojure
(defn foo [a b] (* a b))
```
try placing the cursor before/after each paren and press delete/backspace.
I would expect the editor to comply, even if it would bring the code in a non-symmetrical state.

Note: if I select the parens as text and press del/backspace, it **is** working correctly

Also, copy-paste the following illegal code:
```clojure
defn foo [a b] (* a b))
```
and try to insert a **(** in front, to make it legal.

I suppose this is considered a feature (pressing backspace going to the previous block), but it's not what Idea does by default. Please consider fixing it :)

**Comment by dmarjenburgh** [comment]
Hi Pap,

It's called structural editing (or paredit). It limits the ways in which
you can delete parentheses to prevent creating unbalanced expressions. It
also provides a slew of other functions to manipulate the forms in various
ways.
https://cursiveclojure.com/userguide/paredit.html
http://danmidwood.com/content/2014/11/21/animated-paredit.html

If you are completely new to Clojure/Lisp I understand wanting to turn it
off. After a while though you'd want to master structural editing.
It takes some time and practice, but afterwards you'll notice the amazing
productivity boost you get from it and you don't want to do without it.

**Comment by dmarjenburgh** [comment]
On a related note: It would be nice if structural editing would suppress itself when delimiters are unbalanced so the code can be fixed easily. When everything becomes balanced, it can "activate" again.

**Comment by paplorinc** [comment]
Hi,

I edited my original post, I think you might have missed that part.

Idea for Java has this option also, to insert matching parens or quotes (even for selected text, here it just overrides everything), but my problem here was with fixing wrong formats and changing existing ones.

Not doing anything when I want to close a paren is not an *amazing productivity boost*, it's a very annoying bug, it gives the feeling that the editor wants to be smarter than you :).

**Comment by cursiveclojure** [comment]
@dmarjenburgh Yes, this would be a nice feature. Cursive does actually do this (allows you to delete unbalanced parens) in fairly limited circumstances, but doing it properly is quite tricky. I'll see if I can turn it off automatically, but I suspect that for newcomers, having it switch on and off at apparently random moments is going to be even more confusing.

@paplorinc Intuitive is a relative thing - to someone who's used to structural editing, what you're talking about is actually intuitive and useful. For cases when you need to modify something, you can switch it off as mentioned in the doc.

**Comment by paplorinc** [comment]
By intuitive I mean compared to other IDEs, or at least to the one Cursive resides in: Idea.
New users shouldn't have to read documentation, they should be able to discover it and keep what they like. Forcing them to adopt the One True Method is not user-friendly, in my opinion.

All I'm asking for here is consistency with the host platform's native behavior.

**Comment by cursiveclojure** [comment]
No-one is forced to do anything - if you don't like structural editing, turn it off.

**Comment by SerCeMan** [comment]
+1 for @paplorinc, it is really difficult to even disable it

**Comment by cursiveclojure** [comment]
@SerCeMan How is it difficult to disable? There's an always-visible widget in the status bar to do it with a single click.

**Comment by SerCeMan** [comment]
@cursiveclojure, a lot of developers don't use widgets, if you know hotkeys all you need is editor. And it is really difficult disable option if you don't know option name

**Comment by cursiveclojure** [comment]
@SerCeMan Searching for "structural" in the Settings pane finds the option to turn it off, and searching for "structural" in the Keymap panel shows the "Toggle structural editing" option.

Of course, it's difficult to find that option if you have no idea what you're looking for, but that is also true of any other setting in IntelliJ. The setting is with all the other related settings under *Editor&rarr;General&rarr;Smart Keys* which seems like the first place to look for a problem like this.
